Replace bp/sp-vars by generic stack-based ones

thestr4ng3r opened the pull request:
Pending updates in other projects to be merged after this:
https://github.com/rizinorg/rz-ghidra/pull/307
https://github.com/rizinorg/rz-retdec/pull/16
https://github.com/rizinorg/cutter/pull/3062
https://github.com/rizinorg/jsdec/pull/26

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**
Variables on the stack are not identified by bp/sp+offset anymore, but
by their address from the bottom of the stack frame (RzStackAddr),
independent of how they are accessed.
So now there are only two kinds of variables: stack and register.

This required some major refactoring and other changes:
* RzAnalysisVar.isarg was removed. Whether a variable is an argument is
  now specified implicitly by its storage location.
* Varsub of struct fields had to be rewritten so fields can be queried
  by arbitrary stack addresses using the recently introduced sp
  tracking, as the old approach to fill a list with all fields would not
  work anymore.
* analysis.vars.stackname was removed, new behavior is more similar to
  this being true before.
* Variables will not be created at stack+0 now, because the return
  address is there. Before, vars were only created sometimes in such
  cases.
* Variables created from bp offsets in x86 are not deleted anymore if
  the function's bp_frame is false (see removed
  rz_analysis_function_delete_vars_by_kind(fcn,
  RZ_ANALYSIS_VAR_KIND_BPV); calls). This may lead to some
  false-positive detected variables. Whether this really is a practical
  issue is yet to be seen. At least there are no meaningful tests that
  are broken by this.
* Applying variables from dwarf needed some fixes for determining the
  correct stack locations of variables in order to write meaningful
  tests. The handling is still not entirely correct for all
  possibilities of dwarf info, but at least the changed/added test cases
  are right and serve as a reference for future changes.
* Projects version 11 is introduced.
* afvb commands have been removed, afvs now handles all stack vars.

The key change is the diff in `RzAnalysisVar` (see in particular `RzAnalysisVarStorage`).

Before:
![Bildschirm­foto 2022-12-25 um 20 35 42](https://user-images.githubusercontent.com/1460997/209480059-f415318b-c3df-4af1-85e8-0b3e613e96ed.png)

After:
![Bildschirm­foto 2022-12-25 um 20 36 24](https://user-images.githubusercontent.com/1460997/209480060-bc833c11-1bb8-43aa-b17a-e12da5501058.png)

ret2libc commented on changenotes.txt:
```diff
@@ -0,0 +1,10 @@
+analysis.vars.stackname removed, new behavior is more similar to this being true
```
Is this a temporary file?

thestr4ng3r replied:
Yes, just for my personal notes while working.

wargio commented on librz/include/rz_analysis.h:
```diff
@@ -614,19 +634,39 @@ typedef enum {
+	 * Delta added to register when the var is accessed,
+	 * For example for `[rbp - 0x10]`, this will be -0x10.
+	 */
+	st64 addend;
```
maybe `relative_offset` is a better name ?

thestr4ng3r replied:
I chose this name to deliberately express that is is a value that is added to the register, in direct contrast to the stack offset that is saved in the variable itself and describes the offset from the bottom of the stack.

wargio replied:
i still don't like it, but makes sense.

thestr4ng3r committed:
Replace bp/sp-vars by generic stack-based ones

Variables on the stack are not identified by bp/sp+<offset> anymore, but
by their address from the bottom of the stack frame (RzStackAddr),
independent of how they are accessed.
So now there are only two kinds of variables: stack and register.

This required some major refactoring and other changes:
* RzAnalysisVar.isarg was removed. Whether a variable is an argument is
  now specified implicitly by its storage location.
* Varsub of struct fields had to be rewritten so fields can be queried
  by arbitrary stack addresses using the recently introduced sp
  tracking, as the old approach to fill a list with all fields would not
  work anymore.
* analysis.vars.stackname was removed, new behavior is more similar to
  this being true before.
* Variables will not be created at stack+0 now, because the return
  address is there. Before, vars were only created sometimes in such
  cases.
* Variables created from bp offsets in x86 are not deleted anymore if
  the function's bp_frame is false (see removed
  rz_analysis_function_delete_vars_by_kind(fcn,
  RZ_ANALYSIS_VAR_KIND_BPV); calls). This may lead to some
  false-positive detected variables. Whether this really is a practical
  issue is yet to be seen. At least there are no meaningful tests that
  are broken by this.
* Applying variables from dwarf needed some fixes for determining the
  correct stack locations of variables in order to write meaningful
  tests. The handling is still not entirely correct for all
  possibilities of dwarf info, but at least the changed/added test cases
  are right and serve as a reference for future changes.
* Projects version 11 is introduced.
* afvb commands have been removed, afvs now handles all stack vars.

ret2libc commented on librz/include/rz_analysis.h:
```diff
@@ -600,11 +599,30 @@ typedef struct rz_analysis_bind_t {
+/**
+ * \brief An address on the stack
+ *
+ * These addresses are a relative offset to the value of the stack pointer
+ * when entering a function, i.e. just before the first instruction inside the function is executed.
+ * Thus, with the most common calling conventions, the return address with be at addr 0,
+ * arguments at positive addresses and local variables at negative addresses:
+ *
```
So it means that if I have something like:
```
push r15
push r14
push rbx
sub rsp, 0x50
mov qword [rsp + 10], 0x22
```

the variable at `rsp+10` will be marked as *-88*, right? 8(r15) + 8 (r14) + 8(rbx) + 0x50 - 0x10 == 88 == 0x58.

Did I get it right?

thestr4ng3r replied:
Exactly.

XVilka commented:
```
[XX] C:\projects\rizin\test\db\archos\windows-x64\dbg_rebase var rebase
ANSICON=1 RZ_NOPLUGINS=1 C:\projects\rizin\rizin-vs2022_64-v0.5.0\bin\rizin.exe -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -d -Qc 'aa
afv @ entry0
doc
afv @ entry0
' bins/pe/hello_world/hello_world_od64bit.exe
-- stdout
--- expected
+++ actual
@@ -3,4 +3,4 @@
 var int64_t var_10h @ stack + 0x10
 var int64_t var_18h @ stack - 0x18
 var int64_t var_8h @ stack + 0x8
-var int64_t var_10h @ stack + 0x10EOF
+var int64_t var_10h @ stack + 0x10
-- stderr
Spawned new process with pid 1196, tid = 2000
[x] Analyze all flags starting with sym. and entry0 (aa)
```

XVilka commented on librz/analysis/dwarf_process.c:
```diff
@@ -1700,16 +1713,22 @@ RZ_API void rz_analysis_dwarf_integrate_functions(RzAnalysis *analysis, RzFlag *
+				RzAnalysisVarStorage stor;
+				rz_analysis_var_storage_init_reg(&stor, extra);
+				rz_analysis_function_set_var(fcn, &stor, ttype, 4, var_name);
+			} else if (fcn) { /* kind == 'b' || kind == 's' || kind == 'c' */
```
Maybe better to convert this into a textual comment instead?

XVilka commented on librz/analysis/dwarf_process.c:
```diff
@@ -1700,16 +1713,22 @@ RZ_API void rz_analysis_dwarf_integrate_functions(RzAnalysis *analysis, RzFlag *
+			} else if (fcn) { /* kind == 'b' || kind == 's' || kind == 'c' */
+				RzAnalysisVarStorage stor;
+				RzStackAddr addr = offset;
+				if (*kind == 'b') {
```
What about using an enum here?

thestr4ng3r replied:
Could be done, but this comes from sdb and should rather be refactored completely to not use sdb at all.

XVilka commented on librz/analysis/var.c:
```diff
@@ -8,10 +9,30 @@
+}
+
+/**
+ * Ensure that the register name in \p store comes from the const pool
```
Mismatch between actual parameter name and here

XVilka commented on librz/analysis/var.c:
```diff
@@ -91,49 +87,31 @@ RZ_API void rz_analysis_var_resolve_overlaps(RzAnalysisVar *var) {
 }
 
-RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_set_var(RzAnalysisFunction *fcn, int delta, char kind, RZ_BORROW RZ_NULLABLE const RzType *type, int size, bool isarg, RZ_NONNULL const char *name) {
+RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_set_var(RzAnalysisFunction *fcn, RZ_NONNULL RzAnalysisVarStorage *stor, RZ_BORROW RZ_NULLABLE const RzType *type, int size, RZ_NONNULL const char *name) {
```
Please add Doxygen here

XVilka commented on librz/analysis/var.c:
```diff
@@ -201,12 +175,12 @@ RZ_API void rz_analysis_var_delete(RzAnalysisVar *var) {
 }
 
-RZ_API void rz_analysis_function_delete_vars_by_kind(RzAnalysisFunction *fcn, RzAnalysisVarKind kind) {
+RZ_API void rz_analysis_function_delete_vars_by_storage_type(RzAnalysisFunction *fcn, RzAnalysisVarStorageType stor) {
```
Doxygen.

XVilka commented on librz/analysis/var.c:
```diff
@@ -254,89 +228,179 @@ RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_get_var_byname(RzAnalysisFu
 }
 
-RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_get_var(RzAnalysisFunction *fcn, char kind, int delta) {
+RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_get_var_at(RzAnalysisFunction *fcn, RZ_NONNULL RzAnalysisVarStorage *stor) {
```
Doxygen

XVilka commented on librz/analysis/var.c:
```diff
@@ -254,89 +228,179 @@ RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_get_var_byname(RzAnalysisFu
 	return NULL;
 }
 
+RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_get_stack_var_at(RzAnalysisFunction *fcn, RzStackAddr stack_off) {
```
Doxygen

XVilka replied:
Also `rz_return_if_fail(fcn)`.

XVilka commented on librz/analysis/var.c:
```diff
@@ -254,89 +228,179 @@ RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_get_var_byname(RzAnalysisFu
+	return best;
+}
+
+RZ_API RZ_BORROW RzAnalysisVar *rz_analysis_function_get_reg_var_at(RzAnalysisFunction *fcn, const char *reg) {
```
`RZ_NONNULL` for `reg`? And `rz_return_if_fail(fcn & reg)`

XVilka commented on librz/analysis/var.c:
```diff
@@ -1206,45 +1180,43 @@ RZ_API RZ_OWN RzList /*<RzAnalysisVar *>*/ *rz_analysis_var_list(RZ_NONNULL RzAn
 	return (a && b) ? (a->argnum > b->argnum) - (a->argnum < b->argnum) : 0;
 }
 
 RZ_API void rz_analysis_fcn_vars_cache_init(RzAnalysis *analysis, RzAnalysisFcnVarsCache *cache, RzAnalysisFunction *fcn) {
```
Doxygen

thestr4ng3r replied:
I will deprecate it. This does not make much sense anymore.

XVilka commented on librz/core/canalysis.c:
```diff
@@ -5100,55 +5079,36 @@ RZ_IPI char *rz_core_analysis_var_display(RzCore *core, RzAnalysisVar *var, bool
 		char *r;
 		if (usePxr) {
-			r = rz_core_cmd_strf(core, "pxr $w @ %s%c0x%x", analysis->reg->name[RZ_REG_NAME_BP], sign, udelta);
+			r = rz_core_cmd_strf(core, "pxr $w @r:%s", var->storage.reg);
```
Please add here `// TODO: Convert to the API`

XVilka commented on librz/core/canalysis.c:
```diff
@@ -5100,55 +5079,36 @@ RZ_IPI char *rz_core_analysis_var_display(RzCore *core, RzAnalysisVar *var, bool
+			r = rz_core_cmd_strf(core, "pxr $w @r:%s", var->storage.reg);
 		} else {
-			r = rz_core_cmd_strf(core, "pf %s @ %s%c0x%x", fmt, analysis->reg->name[RZ_REG_NAME_BP], sign, udelta);
+			r = rz_core_cmd_strf(core, "pf r (%s)", var->storage.reg);
```
And here

XVilka commented on librz/core/canalysis.c:
```diff
@@ -5100,55 +5079,36 @@ RZ_IPI char *rz_core_analysis_var_display(RzCore *core, RzAnalysisVar *var, bool
 		char *r;
 		if (usePxr) {
-			r = rz_core_cmd_strf(core, "pxr $w @ %s+0x%x", analysis->reg->name[RZ_REG_NAME_SP], udelta);
+			r = rz_core_cmd_strf(core, "pxr $w @ 0x%" PFMT64x, addr);
```
And here

XVilka commented on librz/core/canalysis.c:
```diff
@@ -5100,55 +5079,36 @@ RZ_IPI char *rz_core_analysis_var_display(RzCore *core, RzAnalysisVar *var, bool
+			r = rz_core_cmd_strf(core, "pxr $w @ 0x%" PFMT64x, addr);
 		} else {
-			r = rz_core_cmd_strf(core, "pf %s @ %s+0x%x", fmt, analysis->reg->name[RZ_REG_NAME_SP], udelta);
+			r = rz_core_cmd_strf(core, "pf %s @ 0x%" PFMT64x, fmt, addr);
```
And here

XVilka commented on librz/core/cmd/cmd_analysis.c:
```diff
@@ -2627,29 +2590,29 @@ RZ_IPI RzCmdStatus rz_analysis_function_vars_handler(RzCore *core, int argc, con
 		break;
 	case RZ_OUTPUT_MODE_RIZIN: {
 		bp = rz_reg_get_name(core->analysis->reg, RZ_REG_NAME_BP);
 		rz_cons_printf("f-fcnvar*\n");
```
Please fix this too while you are on it. `"f- fcnvar*"`

XVilka commented on librz/core/cmd/cmd_analysis.c:
```diff
@@ -2918,83 +2872,96 @@ static RzCmdStatus analysis_function_vars_del(RzCore *core, RzAnalysisVarKind ki
-		delta -= fcn->maxstack;
+	st64 addend = 0;
+	if (stor->type == RZ_ANALYSIS_VAR_STORAGE_STACK) {
+		// TODO: this is wrong. What we need is not the address on the stack, but
```
Probably `FIXME` not `TODO` then

thestr4ng3r replied:
Did we ever specify any difference between `FIXME` and `TODO`? I personally always write `TODO`.

XVilka replied:
Not really. Not that important anyway.

XVilka commented on librz/parse/p/parse_arm_pseudo.c:
```diff
@@ -205,71 +205,117 @@ static bool parse(RzParse *p, const char *assembly, RzStrBuf *sb) {
+		re_str = "\\[([a-z][0-9a-z][0-9]?)\\s*(\\+|-)\\s*(-?(0x)?[0-9a-f]+)\\]";
+		group_idx_sign = 2;
+		group_idx_addend = 3;
+	} else if (op && op->type == RZ_ANALYSIS_OP_TYPE_ADD && op->dst && op->dst->reg && op->dst->reg->name && rz_str_casecmp(op->dst->reg->name, "fp") && rz_str_casecmp(op->dst->reg->name, "sp")) {
```
Maybe split this or extract?

XVilka commented on librz/parse/p/parse_mips_pseudo.c:
```diff
@@ -141,110 +141,103 @@ static bool parse(RzParse *parse, const char *assembly, RzStrBuf *sb) {
+		group_idx_addend = 1;
+	} else {
+		// match e.g. rbp - 0x42
+		// capturing "rbp", "-", "0x42"
```
What `rbp` does in MIPS?

XVilka commented on librz/parse/p/parse_mips_pseudo.c:
```diff
@@ -141,110 +141,103 @@ static bool parse(RzParse *parse, const char *assembly, RzStrBuf *sb) {
+		return tstr;
+	}
+
+	// replace!
```
Replace with what?

XVilka commented on librz/util/json_parser.c:
```diff
@@ -540,22 +540,30 @@ static void json_pj_recurse(const RzJson *json, PJ *pj, bool with_key) {
+ * \param pj the PJ to print into
+ * \param with_key whether to include the root key of \p json, i.e. `"key": <val>`, vs. `<val>`
+ */
+RZ_API void rz_json_to_pj(const RzJson *json, PJ *pj, bool with_key) {
```
Please add `RZ_NONNULL`
